Doc/library/typing.rst

```diff
@@ -1230,6 +1232,83 @@ These are not used in annotations. They are building blocks for creating generic
     ``covariant=True`` or ``contravariant=True``.  See :pep:`484` for more
     details.  By default, type variables are invariant.
 
+.. class:: TypeVarTuple
+
+    :class:`Type variable <TypeVar>` tuple.
+
+    Usage::
+
+        Ts = TypeVarTuple('Ts')
+        Shape = TypeVarTuple('Shape')
+
+    Vanilla type variables enable parameterization with a *fixed* number of
+    types. Type variable tuples, on the other hand, allow parameterization with
+    an *arbitrary* number of types by acting like an *arbitrary* number of type
+    variables wrapped in a tuple. For example::
+
+        class Array(Generic[*Shape]):
+            pass
+
+        from typing import NewType
+        Height = NewType('Height', int)
+        Width = NewType('Width', int)
+
+        1d_array: Array[Height] = Array()  # This is fine
+
+    With one type argument, ``Array`` behaves as if ``Shape = (T,)``, where
+    ``T`` is a type variable. That means ``Array`` behaves as if we'd written
+    ``Generic[T]``. Note the use of the unpacking operator ``*`` to bring ``T``
+    out of the tuple and directly into the square brackets.
```

Comment:
To me, the first two sentences ("With one ... as if we'd written `Generic[T]`.") don't put me on the right track -- there are too many "behaviors" I could imagine that aren't what you're referring to. I personally learn best from examples, so maybe you can show an example of observable behavior for 1d_array? (Note: that's not a valid identifier. :-) I guess that's where my idea of showing `__getitem__` comes in.

The last sentence is useful but maybe could be moved up so it immediately follows the example?

Reply:
> To me, the first two sentences ("With one ... as if we'd written Generic[T].") don't put me on the right track -- there are too many "behaviors" I could imagine that aren't what you're referring to. I personally learn best from examples, so maybe you can show an example of observable behavior for 1d_array?

Fair. I think the primary observable behaviour we want to show here is that `Array` can take an arbitrary number of type arguments, so I've consolidated the `array` instances in the very first example.

> 1d_array? Note: that's not a valid identifier. :-)

Hah, oops :) Fixed.